### ISSUE TITLE
Process

### ISSUE THREAD

**Issue by lillian-lemmer** [hypatia-engine/hypatia#87]
Let's talk about process. Should we require group discussion, requirement docs, architecture planning, before even touching code? I think so.

@ejmr what's your opinion on all this? Any particular references for us? Any particulars to the process of conceptualizing to code process?

**Comment by xeonguy** [comment]
Writing things down helps people to remember and understand things (because they must think slowly about them and explain them). Documentation is the standard way to get people started with any project on their way to coding, for good reason. I personally would not want anyone to touch code they did not understand.

Suggestion: write the skeleton of your architecture document, then delegate each section to be filled out by an aspiring developer of your choosing. Have them document a module and become experts in it.

**Comment by ejmr** [comment]
> Should we require group discussion, requirement docs, architecture planning, before even touching code?

Yes and no.  The most important consideration inherent in these questions is the fact that Hypatia is an open-source project.  Requiring architecture planning and group discussions first is *currently* managable because the team around Hypatia is small enough for everyone to easily communicate with each other.  But once an open-source project explodes in terms of popularity you *must* confront the issue of how to deal with ‘drive-by patches’, i.e. small, one-or-two–commit contributions from people who are never going to work on the project long-term.  The requirements you create, in terms of the upfront investment necessary to contribute code at all, will have a severe effect on the quantity (and quality) of contributions you receive at all.

A game engine is the type of software which eventually grows to the point where *no one* understands every aspect of the architecture, and it becomes unreasonable to expect anyone to understand it.  For example, Linus Torvalds doesn’t have the architecture of the Linux kernel memorized; he delegates the discussion, planning, implementation, etc., of various parts of the kernel to people he trusts who do understand that part of the architecture *because they focus on it exclusively.*  In the long-run Hypatia will be in a similar situation, where you can no longer remember every detail of the engine simply due to its scope.  For example, the game I’ve been working on the past few years is roughly 37,000 lines-of-code in size; do you think I know how everything in that game works?  *Hell no,* heh.

If you require discussion, documentation, planning, etc., from people before accepting their code then you are going to drive away people who would have otherwise contributed.  At the same time though, you obviously don’t want to wildly accept contributions from everyone because then the quality of the project spirals out of control rapidly.  Personally I think the best way to deal with this for open-source software is pretty much what @xeonguy said: delegate parts of the architecture to team members you trust, and demand from *them* discussion, documentation, and planning.  But do not extend those requirements any farther ‘down the chain’.  For example, if you put someone in charge of the animation code, let them accept code contributions of any quality they see fit, from whomever they want.  But when they present the changes to *you,* that’s when you demand a reasonable discussion, plan, justification, so on and so on.

**Comment by xeonguy** [comment]
There is also the situation of phantom coders (where someone forks the codebase, then sends you a pull request after they made their changes). During this time is a great opportunity to ask them a favor before merging, i.e: "Thanks for the contrib. Since you made a code change you must really understand the file. Would it be too much trouble to write the documentation here <link them to the wiki> for the file you changed, before I merge in your code changes? "

Suggestion for good documentation: Have them explain the file, as they understand it, to arbitrary person. IMO, documentation should teach others instead of being notes to oneself.

**Comment by ejmr** [comment]
> Suggestion for good documentation: Have them explain the file, as they understand it, to arbitrary person. IMO, documentation should teach others instead of being notes to oneself.

Agreed, one-hundred percent.  This is the best time, while Hypatia is still relatively in its infancy, to set a high standard for the quality of documentation.  I would follow @xeonguy's suggestion and politely ask (i.e. demand, heh) contributors to document their changes and to document them well.

Personally I also consider commit messages an aspect of documentation, and so I would also recommend setting a standard by-example for quality commit messages.  Ideally anyone should be able to run `git log` and understand *why* any given change was made.  Which means rejecting commits that simply say things like `Fixes a bug` or `Misc updates`, etc.

**Comment by ejmr** [comment]
:+1: on `f321a0cec4a7c9a9efda8a6e46f620d222c82773`.

**Comment by lillian-lemmer** [comment]
Does everyone feel `CONTRIBUTING.md` covers the ideas expressed herein accurately and completely? Are we happy with its contents and thoroughness?
